### ISSUE TITLE
use utf8mb4 encoding as a default

### ISSUE THREAD

**Issue by pangwa** [naushadh/persistent#5]
utf8mb4 supports a wider range characters than utf8(e.g. emoj). It would be safe to use utf8mb4 instead of utf8.

**Comment by naushadh** [comment]
I'd love to move to full UTF-8 range support as well (via `utf8mb4`); however, [upstream commentary #679](https://github.com/yesodweb/persistent/issues/679) suggests this change might be b/w incompatible. ~~The user has no way to "go back" to `utf8` if they want to.~~

Edit: realized something, see below.

**Comment by naushadh** [comment]
@pangwa Can't [`setMySQLConnectInfoCharset`](https://hackage.haskell.org/package/persistent-mysql-haskell-0.4.0/docs/Database-Persist-MySQL.html#v:setMySQLConnectInfoCharset) be used to override the default encoding already?

**Comment by pangwa** [comment]
ah, it looks like setMySQLConnectInfoCharset should work. Thanks. I'm closing this issue.

**Comment by alexkazik** [comment]
There is another issue with utf8/utf8mb4: the create table instruction also uses uft8, and there is no switch for that.
A solution is to specify the sql-type for all textual columns by hand
since it does not help when the connection supports 4-byte-utf8 but the database only 3-byte-utf8.

**Comment by naushadh** [comment]
@alexkazik you're right. re-opening the issue. Just stumbled upon the upstream issue: https://github.com/yesodweb/persistent/issues/813

**Comment by naushadh** [comment]
Could `MySQLConnectInfo`'s charset property be re-used to ensure the migrations are invoked with the right encoding?

**Comment by alexkazik** [comment]
I like the idea, i'd say use utf8mb4 when it's also set in the ci, and urf8(mb3) otherwise - for compatibility.
I'll make a patch for persistent-mysql first and if it's accepted there then also for persistent-mysql-haskell.
Sounds good?

**Comment by naushadh** [comment]
👍
